cmd/registry/cmd/label.go

```diff
@@ -29,223 +29,262 @@ import (
+func updateMap(m map[string]string) (map[string]string, error) {
+	if m == nil {
+		m = make(map[string]string)
+	}
+	if !keyOverwrite {
+		for k, _ := range keysToSet {
+			if v, ok := m[k]; ok {
+				return nil, fmt.Errorf("%q already has a value (%s), and --overwrite is false", k, v)
 			}
 		}
+	}
+	for _, k := range keysToClear {
+		delete(m, k)
+	}
+	for k, v := range keysToSet {
+		m[k] = v
+	}
+	return m, nil
+}
 
-		err = matchAndHandleLabelCmd(ctx, client, taskQueue, args[0])
-		if err != nil {
-			log.Fatalf("%s", err.Error())
-		}
+func init() {
+	var labelCmd = labelOrAnnotateCommand(true)
+	rootCmd.AddCommand(labelCmd)
+	labelCmd.Flags().StringVar(&keyFilter, "filter", "", "Filter selected resources")
+	labelCmd.Flags().BoolVar(&keyOverwrite, "overwrite", false, "Overwrite existing labels")
 
-		close(taskQueue)
-		core.WaitGroup().Wait()
-	},
+	var annotateCmd = labelOrAnnotateCommand(false)
+	rootCmd.AddCommand(annotateCmd)
+	annotateCmd.Flags().StringVar(&keyFilter, "filter", "", "Filter selected resources")
+	annotateCmd.Flags().BoolVar(&keyOverwrite, "overwrite", false, "Overwrite existing annotations")
 }
 
-type labelApiTask struct {
-	ctx    context.Context
-	client connection.Client
-	api    *rpc.Api
-}
+func labelOrAnnotateCommand(commandIsLabel bool) *cobra.Command {
+	return &cobra.Command{
+		Use:   fmt.Sprintf("%s RESOURCE KEY_1=VAL_1 ... KEY_N=VAL_N", commandName()),
+		Short: fmt.Sprintf("%s resources in the API Registry", strings.Title(commandName())),
+		Args:  cobra.MinimumNArgs(2),
+		Run: func(cmd *cobra.Command, args []string) {
+			ctx := context.TODO()
+			client, err := connection.NewClient(ctx)
+			if err != nil {
+				log.Fatalf("%s", err.Error())
+			}
 
-type labelVersionTask struct {
-	ctx     context.Context
-	client  connection.Client
-	version *rpc.ApiVersion
+			taskQueue := make(chan core.Task, 1024)
+			workerCount := 64
+			for i := 0; i < workerCount; i++ {
+				core.WaitGroup().Add(1)
+				go core.Worker(ctx, taskQueue)
+			}
+
+			keysToClear = make([]string, 0)
+			keysToSet = make(map[string]string)
+			for _, operation := range args[1:] {
+				if len(operation) > 1 && strings.HasSuffix(operation, "-") {
+					keysToClear = append(keysToClear, strings.TrimSuffix(operation, "-"))
+				} else {
+					pair := strings.Split(operation, "=")
+					if len(pair) != 2 {
+						log.Fatalf("%q must have the form \"key=value\" (value can be empty) or \"key-\" (to remove the key)", operation)
+					}
+					if pair[0] == "" {
+						log.Fatalf("%q is invalid because it specifies an empty key", operation)
+					}
+					keysToSet[pair[0]] = pair[1]
+				}
+			}
+
+			err = matchAndHandleLabelOrAnnotateCmd(ctx, client, taskQueue, args[0])
+			if err != nil {
+				log.Fatalf("%s", err.Error())
+			}
+
+			close(taskQueue)
+			core.WaitGroup().Wait()
+		},
+	}
 }
 
-type labelSpecTask struct {
+type labelOrAnnotateApiTask struct {
```

Comment:
Consider having separate tasks for labelling and annotating, or having a generic "updateApiTask" type. This particular task feels "specifically generic" where we don't get the benefits of simplicity or reusability. 

Is there a purpose for making this handle both labelling and annotating other than code reuse? The amount of helper functions and reliance on global state required to have it handle two behaviors makes me think we're taking two steps back for one step forward.

Reply:
I didn't like the replication but have split these into handler-specific tasks along with the command implementations.